modules/dnn/src/cuda4dnn/csl/cublas.cpp

```diff
@@ -0,0 +1,313 @@
+
+#include "cublas.hpp"
+#include "stream.hpp"
+#include "pointer.hpp"
+
+#include <cublas_v2.h>
+
+#include <cstddef>
+#include <memory>
+
+#define CUDA4DNN_CHECK_CUBLAS(call) \
+    ::cv::dnn::cuda4dnn::csl::cublas::check((call), CV_Func, __FILE__, __LINE__)
+
+namespace cv { namespace dnn { namespace cuda4dnn { namespace csl { namespace cublas {
+
+    static void check(cublasStatus_t status, const char* func, const char* file, int line) {
+        auto cublasGetErrorString = [](cublasStatus_t err) {
+            switch (err) {
+            case CUBLAS_STATUS_SUCCESS: return "CUBLAS_STATUS_SUCCESS";
+            case CUBLAS_STATUS_NOT_INITIALIZED: return "CUBLAS_STATUS_NOT_INITIALIZED";
+            case CUBLAS_STATUS_ALLOC_FAILED: return "CUBLAS_STATUS_ALLOC_FAILED";
+            case CUBLAS_STATUS_INVALID_VALUE: return "CUBLAS_STATUS_INVALID_VALUE";
+            case CUBLAS_STATUS_ARCH_MISMATCH: return "CUBLAS_STATUS_ARCH_MISMATCH";
+            case CUBLAS_STATUS_MAPPING_ERROR: return "CUBLAS_STATUS_MAPPING_ERROR";
+            case CUBLAS_STATUS_EXECUTION_FAILED: return "CUBLAS_STATUS_EXECUTION_FAILED";
+            case CUBLAS_STATUS_INTERNAL_ERROR: return "CUBLAS_STATUS_INTERNAL_ERROR";
+            case CUBLAS_STATUS_NOT_SUPPORTED: return "CUBLAS_STATUS_NOT_SUPPORTED";
+            case CUBLAS_STATUS_LICENSE_ERROR: return "CUBLAS_STATUS_LICENSE_ERROR";
+            }
+            return "UNKNOWN_CUBLAS_ERROR";
+        };
+
+        if (status != CUBLAS_STATUS_SUCCESS)
+            throw cuBLASException(Error::GpuApiCallError, cublasGetErrorString(status), func, file, line);
+    }
+
+    /** noncopyable cuBLAS smart handle
+     *
+     * UniqueHandle is a smart non-sharable wrapper for cuBLAS handle which ensures that the handle
+     * is destroyed after use. The handle can be associated with a CUDA stream by specifying the
+     * stream during construction. By default, the handle is associated with the default stream.
+     */
+    class Handle::UniqueHandle {
+    public:
+        UniqueHandle() { CUDA4DNN_CHECK_CUBLAS(cublasCreate(&handle)); }
+        UniqueHandle(UniqueHandle&) = delete;
+        UniqueHandle(UniqueHandle&& other) noexcept
+            : stream(std::move(other.stream)), handle{ other.handle } {
+            other.handle = nullptr;
+        }
+
+        UniqueHandle(Stream strm) : stream(std::move(strm)) {
+            CUDA4DNN_CHECK_CUBLAS(cublasCreate(&handle));
+            try {
+                CUDA4DNN_CHECK_CUBLAS(cublasSetStream(handle, StreamAccessor::get(stream)));
+            } catch (...) {
+                /* cublasDestroy won't throw if a valid handle is passed */
+                CUDA4DNN_CHECK_CUBLAS(cublasDestroy(handle));
+                throw;
+            }
+        }
+
+        ~UniqueHandle() noexcept {
+            if (handle != nullptr) {
+                /* cublasDestroy won't throw if a valid handle is passed */
+                CUDA4DNN_CHECK_CUBLAS(cublasDestroy(handle));
```

Comment:
There is a possibility that a previous error could be reported here which would otherwise be an API call which would never throw.

In general, it's annoying since it does not allow exception specifications to be documented properly. As in this case, the destructor is actually `noexcept` but the potential for the API call to throw some unrelated error does not allow marking the destructor `noexcept`. It's even more problematic because a destructor is throwing.

One option is to follow the advice from Effective C++: provide a public destroy method and hope for the user to invoke it; otherwise, eat the exception silently. I don't like this though. It defeats the purpose of even having these RAII based handles.

Another option is to ignore the previous error by removing the `CHECK_CUBLAS` as the documentation guarantees that `cublasDestroy` won't fail for a valid handle. If there was any previous error, it has to be caught later. I like this.

Do you have any opinion on this?

Reply:
Yes, I agree. Don't check the output of `cublasDestory`. It should never fail, except for reporting some other error unrelated to the destroy call, which you can log to STDERR in the destructor and otherwise do nothing.

And as you say, if there really is an error some other CUDA call will blow up in short order elsewhere and the user will get the exception then.